### PR TITLE
Installers using CPack

### PR DESCRIPTION
**Pull request by hsorby**
This work uses CPack to create installers for different OSes.  To make installers run the `package` target.  Different OSes get different installers but all OSes get a compressed archive version for that system.

This hasn't been tested on a Redhat based linux distribution.

Addresses #486.

### CODE REVIEW

File: cmake/common.cmake [hunk 1]
```diff
@@ -197,3 +197,76 @@ function(CONFIGURE_CLANG_AND_CLANG_TIDY_SETTINGS _TARGET)
...
+      if (COMMAND_RESULT EQUAL 0)
+        string(REGEX MATCH "^NAME=\\\"([^\\\"]*)" NAME_OUTPUT ${COMMAND_OUTPUT})
+        string(REPLACE " " "-" SYSTEM_NAME ${CMAKE_MATCH_1})
+      endif()
+    endif()
+    if (NOT SYSTEM_NAME)
+      set(SYSTEM_NAME Unix-Based)
+    endif()
+  else()
+      message( FATAL_ERROR "Don't yet know this system." )
```
Comment (agarny) [inline cmake/common.cmake]: Wrong indentation.

File: cmake/packaging/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  set(CPACK_NSIS_DISPLAY_NAME "${PROJECT_NAME}")
+  set(CPACK_PACKAGE_INSTALL_DIRECTORY "${PROJECT_NAME}")
+  if(CMAKE_SIZEOF_VOID_P EQUAL 4)
+    list(APPEND CPACK_GENERATOR "NSIS")
+  else()
+    list(APPEND CPACK_GENERATOR "NSIS64")
+    set(CPACK_NSIS_INSTALL_ROOT "$PROGRAMFILES64")
+    set(CPACK_NSIS_PACKAGE_NAME "${CPACK_PACKAGE_INSTALL_DIRECTORY} (${LIBCELLML_ARCHITECTURE})")
+    set(CPACK_PACKAGE_INSTALL_REGISTRY_KEY "${CPACK_PACKAGE_NAME} ${CPACK_PACKAGE_VERSION} (${LIBCELLML_ARCHITECTURE})")
+  endif()
```
Comment (agarny) [inline cmake/packaging/CMakeLists.txt:38]: We should check whether NSIS is installed.

File: cmake/packaging/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                                         !define MUI_UNWELCOMEFINISHPAGE_BITMAP \\\"${CMAKE_CURRENT_SOURCE_DIR}/distrib/windows\\\\libcellml_banner_vert.bmp\\\"")
+  set(CPACK_NSIS_MODIFY_PATH "ON")
+elseif(APPLE)
+  # Preserve the CMAKE_INSTALL_PREFIX for the project and work with absolute install
+  set(CPACK_SET_DESTDIR ON)
+
+  set(CPACK_RESOURCE_FILE_WELCOME "${CMAKE_CURRENT_SOURCE_DIR}/distrib/osx/welcome.txt")
+
+  set(LIBCELLML_ARCHITECTURE "universal")
+  set(CPACK_GENERATOR  "TGZ" "productbuild")
```
Comment (agarny) [inline cmake/packaging/CMakeLists.txt]: We should also use `productbuild` and `hdiutil` to generate a `.dmg` file.

File: cmake/packaging/distrib/osx/welcome.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+This installer will guide you through the installation of libCellML on your machine.  The installer will install the libCellML library and API header files.
```
Comment (agarny) [inline cmake/packaging/distrib/osx/welcome.txt]: No double spaces after a full point.

**Comment by agarny** [comment]
We should probably have a test to test the building of our packages?

**Comment by hsorby** [comment]
> We should probably have a test to test the building of our packages?

We probably should.

File: cmake/packaging/distrib/osx/welcome.txt [hunk 5]
```diff
@@ -1,3 +1,3 @@
-This installer will guide you through the installation of libCellML on your machine.  The installer will install the libCellML library and API header files.
+This installer will guide you through the installation of libCellML on your machine. The installer will install the libCellML library and API header files.
 
-libCellML Developers.
+Thanks and enjoy, libCellML Developers.
```
Comment (agarny) [inline cmake/packaging/distrib/osx/welcome.txt]: `developers` rather than `Developers`?

**Review by nickerso — DISMISSED** [review]
This seems to work for me on Windows with the zip and NSIS installers (after I installed NSIS). The NSIS package installed fine and I could even uninstall it. Documentation on supported packagers and installers and their use (to create packages and installers) is required, but I'm happy to defer that to a separate issue.

**Review by agarny — APPROVED** [review]
The banners and icon for the Windows installer look pretty bad. Is that something that Zara could (quickly) improve?

**Comment by agarny** [comment]
I believe this is the next PR that should be merged in, @nickerso, @kerimoyle.

**Comment by nickerso** [comment]
> The banners and icon for the Windows installer look pretty bad. Is that something that Zara could (quickly) improve?

as part of a future issue :) I know @hsorby had some ideas about other updates that could be done for libCellML along these lines.

**Review by nickerso — APPROVED** [review]

**Pull request merged**